Imagine/Filter/PostProcessor/CwebpPostProcessor.php

```diff
@@ -0,0 +1,283 @@
+     * achieved by using a value of **100**.
+     *
+     * In case of lossless compression (specified by the **-lossless** option), a small factor enables faster
+     * compression speed, but produces a larger file. Maximum compression is achieved by using a value of **100**.
+     *
+     *
+     * @var int
+     */
+    protected $q;
+
+    /**
+     * Specify the compression factor for alpha compression between **0** and **100**. Lossless compression of alpha is
+     * achieved using a value of **100**, while the lower values result in a lossy compression.
+     *
+     * @var int
+     */
+    protected $alphaQ;
+
+    /**
+     * Specify the compression method to use. This parameter controls the trade off between encoding speed and the
+     * compressed file size and quality. Possible values range from **0** to **6**. When higher values are used, the
+     * encoder will spend more time inspecting additional encoding possibilities and decide on the quality gain. Lower
+     * value can result in faster processing time at the expense of larger file size and lower compression quality.
+     *
+     * @var int
+     */
+    protected $m;
+
+    /**
+     * Specify the predictive filtering method for the alpha plane. One of **none**, **fast** or **best**, in
+     * increasing complexity and slowness order. Internally, alpha filtering is performed using four possible
+     * predictions (none, horizontal, vertical, gradient). The **best** mode will try each mode in turn and pick the
+     * one which gives the smaller size. The **fast** mode will just try to form an a priori guess without testing all
+     * modes.
+     *
+     * @var string
+     */
+    protected $alphaFilter;
+
+    /**
+     * Specify the algorithm used for alpha compression: **0** or **1**. Algorithm **0** denotes no compression, **1**
+     * uses WebP lossless format for compression.
+     *
+     * @var int
+     */
+    protected $alphaMethod;
+
+    /**
+     * Preserve RGB values in transparent area. The default is off, to help compressibility.
+     *
+     * @var bool
+     */
+    protected $exact;
+
+    /**
+     * An array of metadata to copy from the input to the output if present. Valid values: **all**, **none**, **exif**,
+     * **icc**, **xmp**.
+     *
+     * Note that each input format may not support all combinations.
+     *
+     * @var string[]
+     */
+    protected $metadata;
+
+    /**
+     * @var OptionsResolver
+     */
+    private $resolver;
+
+    public function __construct(
+        string $executablePath = '/usr/bin/cwebp',
+        string $temporaryRootPath = null,
+        int $q = null,
+        int $alphaQ = null,
+        int $m = null,
+        string $alphaFilter = null,
+        int $alphaMethod = null,
+        bool $exact = null,
+        array $metadata = []
+    ) {
+        parent::__construct($executablePath, $temporaryRootPath);
+
+        $this->q = $q;
```

Comment:
my idea was to create the options resolver with the arguments and then set the properties from the resolved options. that way, they are validated and we never assign something invalid to the object properties.

but i realize we need to validate twice, first for the constructor and then with the options specified in a call. what does optionsresolver do when one of the defaults does not match the allowed values? (the allowed types are ensured by the static types on the constructor) if it reports a problem, the current implementation seems good to me.

Reply:
I'm not a big fan of calling the OptionResolver in the constructor, I prefer to keep it simple (even if yes it a part of his job to ensure data are valid). 

But anyway, if the default parameter if wrong, an error will occurred as expected!

Reply:
as long as an invalid default does not make the resolver think that its actually ok to have that wrong value, i'd say we are good. thanks for checking